src/main/java/org/candlepin/controller/EntitlementCertificateGenerator.java

```diff
@@ -334,6 +340,95 @@ public void regenerateCertificatesOf(Consumer consumer, boolean lazy) {
         this.regenerateCertificatesOf(consumer.getEntitlements(), lazy);
     }
 
+    @Transactional
+    public void regenerateCertificatesOf(Consumer consumer, List<String> current,
+        List<String> envIds, boolean lazy) {
+        log.info(
+            "Regenerating #{}, entitlement certificates for consumer: {}",
+            consumer.getEntitlements().size(), consumer
+        );
+
+        // we need to clear the content access cert on regenerate
+        Owner owner = ownerCurator.findOwnerById(consumer.getOwnerId());
+        if (owner.isUsingSimpleContentAccess()) {
+            this.contentAccessManager.removeContentAccessCert(consumer);
+        }
+
+        List<Entitlement> ents = filterEntitlements(consumer, current, envIds);
+
+        // TODO - Assumes only 1 entitlement certificate exists per entitlement
+        this.regenerateCertificatesOf(ents, lazy);
+    }
+
+    private List<Entitlement> filterEntitlements(Consumer consumer, List<String> current,
+        List<String> envList) {
+        List<Entitlement> ents = new ArrayList<>();
+
+        for (Entitlement e : consumer.getEntitlements()) {
+            boolean entToRegen = false;
+
+            for (String env : envList) {
+                List<String> listOfPc = getContentUUIDsFromEntitlement(e);
+                List<String> currentHigher = null;
+                List<String> envContent = this.environmentCurator.getEnvironmentContent(env);
+                List<String> envListByCurrentHigher = new ArrayList<>();
+
+                if (current.contains(env)) {
+                    currentHigher = new ArrayList<>(current.subList(0, current.indexOf(env)));
+                }
+
+                for (String higherEnv : currentHigher) {
+                    envListByCurrentHigher.addAll(this.environmentCurator.getEnvironmentContent(higherEnv));
+                }
+
+                for (String envC : envContent) {
+                    if (listOfPc.contains(envC) && !envListByCurrentHigher.contains(envC)) {
+                        entToRegen = true;
+                        break;
+                    }
+                }
+            }
+
+            if (entToRegen) {
+                ents.add(e);
+            }
+        }
+
+        return ents;
+    }
+
+    private List<String> getContentUUIDsFromEntitlement(Entitlement entitlement) {
+        List<String> listOfPc = new ArrayList<>();
+        // Entitlement content UUIDs
+        Collection<ProductContent> pc = entitlement.getPool().getProduct().getProductContent();
```

Comment:
Given the frequency at which we're calling and using the pool and product, they should be stored locally.

Also, add a line of whitespace between the blocks below.

Reply:
this method is now removed.